File: problem 17/solution.js
var isValid = function(s) {
    const stack = []; // stack for opening brackets
    const hash = { ')': '(', ']': '[', '}': '{' }; // matching pairs
    
    for (const char of s) {
        if (char in hash) {
            // check if stack is non-empty and matches the top element
            if (stack.length && stack[stack.length - 1] === hash[char]) {
                stack.pop(); // remove the matching opening bracket
            } else {
                return false; // invalid if no match
            }
        } else {
            stack.push(char); // push opening bracket
        }
    }
    return stack.length === 0; // valid if stack is empty
};

File: problem 17/solution.cs
using System.Collections.Generic;

public class Solution {
    public bool IsValid(string s) {
        Stack<char> stack = new Stack<char>(); // stack for opening brackets
        
        foreach (char c in s) {
            if (c == '(' || c == '{' || c == '[') {
                stack.Push(c); // push opening bracket
            } else {
                // check if stack is empty or doesn't match
                if (stack.Count == 0 || 
                   (c == ')' && stack.Pop() != '(') || 
                   (c == '}' && stack.Pop() != '{') || 
                   (c == ']' && stack.Pop() != '[')) {
                    return false;
                }
            }
        }
        return stack.Count == 0; // valid if stack is empty
    }
}
